### PR TITLE
Add tests (and fix) cookie fallback mode

### PR DESCRIPTION
**Pull request by kildareflare**
Hi, I was in the process of adding a reset method to this library (clear all local storage / cookies) and noticed that there were no tests for cookie fallback mode. 

I also noticed that the ngCookie dependency was not declared; I may be missing something, but I don't think the cookie fallback mode will work without it.

I've added the missing dependency and written tests for the cookie fallback mode - all tests pass.

The changes for the reset method are ready to be committed, pending the merge of these changes.

This is my first pull request, so please let me know if I have missed something.

Cheers 

image

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -21,7 +21,8 @@
   },
   "devDependencies": {
     "angular-mocks": ">= 1.2",
-    "angular-scenario": ">= 1.2"
+    "angular-scenario": ">= 1.2",
+	"angular-cookies": ">= 1.4"
```
Comment (mgonto) [inline bower.json]: You need to change Tabs to Spaces that's why it's "moved" here

File: dist/angular-storage.js [hunk 2]
```diff
@@ -94,7 +94,7 @@ angular.module('angular-storage.storage', [])
   }]);
 
 
-angular.module('angular-storage.store', ['angular-storage.internalStore'])
+angular.module('angular-storage.store', ['angular-storage.internalStore', 'ngCookies'])
```
Comment (mgonto) [inline dist/angular-storage.js]: Don't add it here as a dependency. If you add it here as a dependency, it means that it MUST be present to start the app. Some people might not want the cookie fallback, and even if they don't, they have to add the `angular-cookies.js` if you add the module as a dependency. That's why, I'm using the `$injector` to get the dependency and not depending on `ngCookies`. Makes sense?

File: src/angularStorage/services/store.js.js [hunk 3]
```diff
@@ -1,4 +1,4 @@
-angular.module('angular-storage.store', ['angular-storage.internalStore'])
+angular.module('angular-storage.store', ['angular-storage.internalStore', 'ngCookies'])
```
Comment (mgonto) [inline src/angularStorage/services/store.js.js]: Same here

File: test/unit/angularStorage/services/storeSpec.js [hunk 4]
```diff
@@ -69,6 +69,100 @@ describe('angularStorage store', function() {
...
+    * note - to confirm that cookiestore was used we attempt to retrieve the value from the cookie
+             since this bypasses our service, the result will not have been json parsed
+             therefore we use JSON.stringify on the expected value, so comparing like for like 
+    *
+    */
+    
+    var windowMock, $cookieStore;
+    
+    /* provide a mock for $window where localStorage is not defined */
+    beforeEach(module("angular-storage.store", function ($provide) {
```
Comment (mgonto) [inline test/unit/angularStorage/services/storeSpec.js]: As we won't make angular-storage depend on `ngCookies`, what we should do is that in the tests, the module that gets created for testing should depend on `ngCookies` which is what would happen if we were doing a real app. Makes sense?

**Comment by mgonto** [comment]
Hey @kildareflare 

Thanks for the contribution! It's a very nice one ;). Congrats on the first PR as well :).

I've put a few changes to the PR. Please do let me know if they make sense and if you can do them before actually merging it. Makes sense?

Thanks again!

**Comment by kildareflare** [comment]
Hey @mgonto 

Yes everything makes sense, I thought it might have been something like that (making the dependency optional) but became blinkered in needed the dependency to run the tests.
I'll try and make the changes tonight; if not I'll make them when I get back from my hols.

Cheers

**Comment by mgonto** [comment]
Hey,

Otherwise, I can continue from your PR :).

Basically for the tests, you can inject 

````js
cookieStore = $injector.get('$cookieStore');
````

**Comment by kildareflare** [comment]
It would be nice to complete myself - I'll aim to do it tonight, but if I don't get a chance, feel free to continue...
PS I've just started work, so tonight is 8 hours away :0)

**Comment by mgonto** [comment]
I can wait for you, no problem :).

Just let me know if you need any further help!

Thanks for contributing!

**Comment by mgonto** [comment]
We're in different timezones :P it's 5PM here

**Comment by kildareflare** [comment]
OK, I think I managed to get all done, off to pack now.
If all good, I'll submit a PR for the other changes when I get back.
Cheers

**Comment by mgonto** [comment]
Now it looks awesome @kildareflare 

Nice work!!

Congrats on your first PR being merged ;) 

Thanks!

**Comment by kildareflare** [comment]
No worries and thanks!

I started work on adding the reset method today and was just about to create the pull request when I realised that it was not that straightforward - we need to be careful that we only remove 'our' items from local storage. Likewise the cookies.

**Comment by mgonto** [comment]
What I'd suggest for that @kildareflare is using a [Namespaced storage](https://github.com/auth0/angular-storage#namespaced-storages).

Once you have it, to delete all values is just to delete all keys from `localStorage` that actually start with the specified prefix. This means you won't have to save all values that were saved by you nor anything like that.

Makes sense?

**Comment by kildareflare** [comment]
Yes, that was what I was thinking - however this only covers the scenario where a namespaced store has been created. If the user creates a store without the namespacing we have no way to distinguish 'our' items / cookies; no?

**Pull request merged**